### PR TITLE
Rename Blockly.<language>.variableDB_ to .nameDB_ and add .nameDB_.getUserNames('VARIABLE') and .nameDB_.getUserNames('PROCEDURE')

### PR DESCRIPTION
**Pull request by NeilFraser**
Should be no user-visible changes.

### CODE REVIEW

**Comment by NeilFraser** [comment]
The net result of all of these commits is that in any block's generator one can now call either one of:
```
Blockly.JavaScript.nameDB_.getUserNames('VARIABLE')
Blockly.JavaScript.nameDB_.getUserNames('PROCEDURE')
```
and get an array of all the user's variable or procedure names.

File: core/generator.js [hunk 1]
```diff
@@ -402,32 +402,50 @@ Blockly.Generator.prototype.definitions_;
...
+Object.defineProperty(Blockly.Generator.prototype, 'variableDB_', {
+  get: function() {
+    console.warn('Deprecated use of "variableDB_"; change to "nameDB_"');
+    return this.nameDB_;
+  },
+  set: function(x) {
+    console.warn('Deprecated use of "variableDB_"; change to "nameDB_"');
+    this.nameDB_ = x;
+  }
+});
```
Comment (BeksOmega) [inline core/generator.js]: This seems like a nice way to deprecate something! I'll have to keep this in mind in the future.

File: core/generator.js [hunk 2]
```diff
@@ -402,32 +397,59 @@ Blockly.Generator.prototype.definitions_;
...
+
+Object.defineProperty(Blockly.Generator.prototype, 'variableDB_', {
+  /**
+   * Getter.
+   * @deprecated 'variableDB_' was renamed to 'nameDB_' (May 2021).
+   * @this {Blockly.Generator}
+   * @return {Blockly.Names} Name database.
+   */
+  get: function() {
+    console.warn('Deprecated use of "variableDB_"; change to "nameDB_"');
```
Comment (maribethb) [inline core/generator.js:414]: We have the Blockly.utils.deprecation.warn function, which would be good to use for consistency and so we can more easily track what we have scheduled for deletion

**Review by maribethb — APPROVED** [review]
It was helpful that the cleanup changes were in separate commits, but it'd be even easier to review if they were in separate PRs. This also matters because your PR title will be in the release notes, and having variable and procedure names available on demand might be something devs are interested in seeing at a glance from the notes.

File: core/generator.js [hunk 3]
```diff
@@ -402,32 +397,59 @@ Blockly.Generator.prototype.definitions_;
...
  */
-Blockly.Generator.prototype.variableDB_;
+Blockly.Generator.prototype.nameDB_;
+
+Object.defineProperty(Blockly.Generator.prototype, 'variableDB_', {
+  /**
+   * Getter.
+   * @deprecated 'variableDB_' was renamed to 'nameDB_' (May 2021).
+   * @this {Blockly.Generator}
+   * @return {Blockly.Names} Name database.
```
[the remark this answers is not in the document]
Comment (NeilFraser) [reply core/generator.js]: Fixed here, and a few dozen other places.

File: core/generator.js (hunk 2 above)
Comment (NeilFraser) [reply core/generator.js]: Fixed, and changed Blockly.utils.deprecation.warn to also handle properties, not just functions.

**Comment by NeilFraser** [comment]
> It was helpful that the cleanup changes were in separate commits, but it'd be even easier to review if they were in separate PRs. This also matters because your PR title will be in the release notes, and having variable and procedure names available on demand might be something devs are interested in seeing at a glance from the notes.

The flip side is that splitting this into multiple PRs would result in one of:
1) A set of mostly serial PRs that each block the next and takes much longer
2) An unmanageable set of merge collisions
3) Pressure to not fix stuff all over the codebase.

I'm concerned that the current process around release notes is resulting in lower quality code since there's a penalty for resolving side issues on the spot upon discovery.  Is there a way we can build informative release notes without discouraging off-topic edits?

**Comment by maribethb** [comment]
> > It was helpful that the cleanup changes were in separate commits, but it'd be even easier to review if they were in separate PRs. This also matters because your PR title will be in the release notes, and having variable and procedure names available on demand might be something devs are interested in seeing at a glance from the notes.
> 
> The flip side is that splitting this into multiple PRs would result in one of:
> 
> 1. A set of mostly serial PRs that each block the next and takes much longer
> 2. An unmanageable set of merge collisions
> 3. Pressure to not fix stuff all over the codebase.
> 
> I'm concerned that the current process around release notes is resulting in lower quality code since there's a penalty for resolving side issues on the spot upon discovery. Is there a way we can build informative release notes without discouraging off-topic edits?

Makes sense. This is one of those times I look longingly over at internal-Google systems and remember how easy it is to have multiple changes going at once. The good ol' days.

I probably would have moved all of the cleanup changes to their own PR, and have a separate PR for the non-cleanup changes around name generation to get that balance between clarity/ease of reviewing and workflow, but I realize everyone sets that threshold differently, so it's not a big deal. 

I do think it's a good idea to update the PR title to mention the generator change though, as it's most relevant from a release notes perspective. I'm not sure how best to improve them, since it needs to be an automatic process, and grabbing every commit title is too granular, imo. We could look into it in our next "non-feature week" though! There's probably some tooling out there for it, or we could have a section in the PR template for release notes and parse that out later.

**Pull request merged**